src/main/java/net/sf/jabref/gui/mergeentries/MergeEntryDOIDialog.java

```diff
@@ -162,21 +169,22 @@ private void buttonPressed(String button) {
 
             // fields
             for (String field : jointFields) {
-                String originalString = originalEntry.getField(field);
-                String mergedString = mergedEntry.getField(field);
-                if ((originalString == null) || !originalString.equals(mergedEntry.getField(field))) {
-                    originalEntry.setField(field, mergedString);
-                    ce.addEdit(new UndoableFieldChange(originalEntry, field, originalString, mergedString));
+                Optional<String> originalString = originalEntry.getFieldOptional(field);
+                Optional<String> mergedString = mergedEntry.getFieldOptional(field);
+                if (!originalString.isPresent() || !originalString.equals(mergedString)) {
+                    originalEntry.setField(field, mergedString.get()); // mergedString always present
+                    ce.addEdit(new UndoableFieldChange(originalEntry, field, originalString.orElse(null),
+                            mergedString.get()));
                     edited = true;
                 }
             }
 
-            // Remove fields which are not in the merged entry
+            // Remove fields which are not in the merged entry, unless they are internal fields
             for (String field : originalFields) {
-                if (!jointFields.contains(field)) {
-                    String originalString = originalEntry.getField(field);
+                if (!jointFields.contains(field) && !field.startsWith("__")) {
```

Comment:
Maybe it would make sense to add a method `public static boolean isInternalField(String field)` to `EntryUtil` and use it for the second part of the condition statement?

Just so that the knowledge on the structure of internal fields moves into the model package.